File: Problems/Algorithms/55. Jump Game/JumpGame.java
class Solution {
    public boolean canJump(int[] nums) {
        int n = nums.length;
        boolean[] dp = new boolean[n];
        dp[n-1] = true;
        
        for (int i = n-2; i >= 0; i--) {
            if (nums[i] >= n-1-i) {
                dp[i] = true;
            } else {
                for (int j = i+1; j <= i+nums[i]; j++) {
                    dp[i] = dp[i] || dp[j];
                }
            }
        }
        
        return dp[0];
    }
}

File: Problems/Algorithms/55. Jump Game/JumpGame.kt
class Solution {
    fun canJump(nums: IntArray): Boolean {
        val n = nums.size
        val dp = BooleanArray(n)
        dp[n-1] = true
        
        for (i in n-2 downTo 0) {
            if (nums[i] >= n-1-i) {
                dp[i] = true
            } else {
                dp[i] = dp.copyOfRange(i+1, i+nums[i]+1).any { it }
            }
        }
        
        return dp[0]
    }
}

